src/Operation/ReplaceOne.php

```diff
@@ -96,6 +103,23 @@ public function __construct(string $databaseName, string $collectionName, $filte
             throw new InvalidArgumentException('$replacement is an update pipeline');
         }
 
+        if (isset($options['codec']) && ! $options['codec'] instanceof DocumentCodec) {
+            throw InvalidArgumentException::invalidType('"codec" option', $options['codec'], DocumentCodec::class);
+        }
+
+        if (isset($options['codec'])) {
+            if (isset($options['typeMap'])) {
+                throw InvalidArgumentException::cannotCombineCodecAndTypeMap();
+            }
+
+            $replacement = $options['codec']->encodeIfSupported($replacement);
```

Comment:
I'll note that this differs from BulkWrite, which calls `encodeIfSupported()` at execution time. I understand that's necessary because we're just delegating to a common Update operation here, but it may be worth a comment.

Also, I think the assertion below can be moved within this `if (isset($options['codec']))` condition.

Reply:
That is correct. When we create the nested `UpdateOne` operation, we'd have to inspect the update document to figure out whether we're doing an atomic update or a replace, so I opted for handling encoding of data in `ReplaceOne`, even though this causes an exception at a different time than in other cases.

I've moved the assertion here as well.